src/runtime.js

```diff
@@ -17,43 +17,64 @@ function replaceComponent(oldType, newType) {
 
 		if (vnode.__c) {
 			vnode.__c.constructor = vnode.type;
-			if (vnode.__c.__H) {
-				// Reset hooks state
-				vnode.__c.__H = {
-					__: [], // _list
-					__h: [] // _pendingEffects
-				};
-			}
 
-			// if this was/is a class component:
-			if (vnode.__c instanceof oldType) {
-				const oldInst = vnode.__c;
-				const newInst = new newType(vnode.__c.props, vnode.__c.context);
-				vnode.__c = newInst;
-				// copy old properties onto the new instance.
-				//   - Objects (including refs) in the new instance are updated with their old values
-				//   - Missing or null properties are restored to their old values
-				//   - Updated Functions are not reverted
-				//   - Scalars are copied
-				for (let i in oldInst) {
-					const type = typeof oldInst[i];
-					if (!(i in newInst)) {
-						newInst[i] = oldInst[i];
-					} else if (type !== 'function' && typeof newInst[i] === type) {
-						if (
-							type === 'object' &&
-							newInst[i] != null &&
-							newInst[i].constructor === oldInst[i].constructor
-						) {
-							Object.assign(newInst[i], oldInst[i]);
-						} else {
+			try {
+				if (vnode.__c instanceof oldType) {
+					const oldInst = vnode.__c;
+					const newInst = new newType(vnode.__c.props, vnode.__c.context);
+					vnode.__c = newInst;
+					// copy old properties onto the new instance.
+					//   - Objects (including refs) in the new instance are updated with their old values
+					//   - Missing or null properties are restored to their old values
+					//   - Updated Functions are not reverted
+					//   - Scalars are copied
+					for (let i in oldInst) {
+						const type = typeof oldInst[i];
+						if (!(i in newInst)) {
 							newInst[i] = oldInst[i];
+						} else if (type !== 'function' && typeof newInst[i] === type) {
+							if (
+								type === 'object' &&
+								newInst[i] != null &&
+								newInst[i].constructor === oldInst[i].constructor
+							) {
+								Object.assign(newInst[i], oldInst[i]);
+							} else {
+								newInst[i] = oldInst[i];
+							}
 						}
 					}
 				}
+			} catch (e) {
+				/* Functional component */
 			}
 
-			Component.prototype.forceUpdate.call(vnode.__c);
+			if (vnode.__c.__H) {
+				const visited = new WeakSet();
+				const hooks = vnode.__c.__H;
+				const oldOptionsHook = options.__h;
+				options.__h = (component, index, type) => {
+					const hooks = component.__H.__;
+					if (hooks[index].type !== type || !hooks[index].__hot_reload_id__) {
+						hooks.splice(index, 0, { type });
+					} else if (hooks[index].__hot_reload_id__) {
+						visited.add(hooks[index].__hot_reload_id__);
+					}
+
+					if (oldOptionsHook) oldOptionsHook(component, index, type);
+				};
+
+				Component.prototype.forceUpdate.call(vnode.__c, () => {
+					vnode.__c.__H.__.forEach(({ __hot_reload_id__: hotReloadId }, i) => {
```

Comment:
Now that we have finished forcefully updating we should check if we have any unvisited children in the hooks array since these could have been removed.

Reply:
Would love for this to be included as a comment in the code :+1: